rfc/rfc-79/rfc-79.md

```diff
@@ -0,0 +1,138 @@
+
+If a single write handle need to write 3 files:
+Say a single write handle needs to write 3 data files due to file sizing overflows. And is crashed after writing just 2 of them. Based on above proposal, on retry, all 3 data files needs to be written out again. Bcoz, the boundaries at which the records get split across data files it not deterministic. Say we have to write 2.5M records. In first attempt, we could be writing 1M in file1, 900k in 2nd file and 600k in 3rd file. On re-attempt, we could be splitting the records as 950k in first file, 900k in 2nd file nad 650k in 3rd file. So, completion marker tracks whether an entire write handle is complete or not, i.e. entire spark task is fully complete or not.
+This also means that, we might be serializing N no of write statuses to the Completion marker.
+
+### 2. Finalize Write markers:
+   Entire spark stage has to be retried after marker based reconciliation is complete(case2). So, lets add a finalize write markers which is just 1 marker per write.
+
+AIs:
+We have two options to go about when such behavior is seen. Either we could fail the entire write or we could re-generate the write status.
+
+Within write handle, if we deduce the finalize write markers:
+If config is enabled:
+Fail the entire write.
+
+If config is not enabled // which means to not throw exception
+Get hold of already written data files, parse them to generate writeStatus (since we need to get hold of hoodie record keys in WriteStatus) and return them.
+
+Even in 2nd case, we do not generate new data files, but just read back the already written files and serve them.
+
+With both the fixes in play, we can completely avoid generating spurious data files. And even if entire stage is re-tried/re-executed, we could avoid a full rewrite of the data files.
+
+
+## Implementation nuances:
+
+We need the implement the aforementioned design in all write handles. But we also need to ensure one more thing apart from generating and acting on the two markers.
+FileID generation:
+We need to ensure we generate fileID prefixes in the driver and not in the executor. Unless we make them deterministic, w/ clustering and create handles, a re-attempt could generate a new fileID and one of the above fix will not catch the spurious data file issue.
+
+- Bulk insert (write client) needs fixes to generate fileID prefixes in driver.
+- Bulk insert in row write code path needs fixes.
+- Inserts and upserts are good. (no fixes required)
+- Clustering under the hood uses bulk insert. So, no additional fixes needed.
+
+Overall: we need to ensure fileID prefixes are generated in the driver once and not re-generated on re-attempts, we should be good. Even if we have file sizing overflow cases, we suffix the indexes to the fileID prefix, and so we should be good. Our completion markers are designed one per fileId prefix and so entire spark task will be retried on any crashes/failures mid-way for a given spark task.
```

Comment:
to consider:  
w/ stray task issue: consider case where completion marker is fully deleted and so we may not be able to guard it.

Reply:
Also another case to consider: 
a write handle might try to delete a data file at the end of writing it. But process might crash and might leave the dangling file on storage.